engine/access/subscription/util.go

```diff
@@ -11,29 +13,50 @@ import (
 // handles the received responses, and sends the processed information to the client via the provided stream using handleResponse.
 //
 // Parameters:
+// - ctx: Context for the operation.
 // - sub: The subscription.
 // - handleResponse: The function responsible for handling the response of the subscribed type.
 //
-// Expected errors during normal operation:
-//   - codes.Internal: If the subscription encounters an error or gets an unexpected response.
-func HandleSubscription[T any](sub Subscription, handleResponse func(resp T) error) error {
+// No errors are expected during normal operations.
+func HandleSubscription[T any](ctx context.Context, sub Subscription, handleResponse func(resp T) error) error {
 	for {
-		v, ok := <-sub.Channel()
-		if !ok {
-			if sub.Err() != nil {
-				return rpc.ConvertError(sub.Err(), "stream encountered an error", codes.Internal)
+		select {
+		case v, ok := <-sub.Channel():
+			if !ok {
+				if sub.Err() != nil {
+					return fmt.Errorf("stream encountered an error: %w", sub.Err())
+				}
+				return nil
 			}
-			return nil
-		}
 
-		resp, ok := v.(T)
-		if !ok {
-			return status.Errorf(codes.Internal, "unexpected response type: %T", v)
-		}
+			resp, ok := v.(T)
+			if !ok {
+				return fmt.Errorf("unexpected response type: %T", v)
+			}
 
-		err := handleResponse(resp)
-		if err != nil {
-			return err
+			err := handleResponse(resp)
+			if err != nil {
+				return err
+			}
+		case <-ctx.Done():
```

Comment:
the api subscribe methods will setup a streamer which will close the subscription's channel when the passed context is canceled. the existing code should handle that already

Reply:
[22ad469](https://github.com/onflow/flow-go/pull/6636/commits/22ad4699dc75b52c6cd2e1682d0efdcabcd9fa0c)